tests/fe/abf_02.cc

```diff
@@ -0,0 +1,593 @@
+
+#include <deal.II/fe/mapping_q.h>
+
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/lac/vector.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/vector_memory.h>
+#include <deal.II/lac/sparsity_pattern.h>
+#include <deal.II/lac/sparse_matrix.h>
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/lac/constraint_matrix.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+
+
+
+
+
+/*
+ * Check the value of the derivative field.
+ */
+
+void EvaluateDerivative (DoFHandler<3> &dof_handler,
+                         Vector<double> &solution)
+{
+  // This quadrature rule determines the points, where the
+  // derivative will be evaluated.
+  QGauss<3> quad (3);
+  FEValues<3> fe_values (dof_handler.get_fe (), quad,
+                         UpdateFlags(update_values    |
+                                     update_quadrature_points  |
+                                     update_gradients |
+                                     update_JxW_values));
+
+  const unsigned int   n_q_points    = quad.size();
+  const unsigned int   n_components   = dof_handler.get_fe().n_components();
+  const unsigned int dofs_per_cell = dof_handler.get_fe().dofs_per_cell;
+
+  // Cell iterators
+  DoFHandler<3>::active_cell_iterator cell = dof_handler.begin_active(),
+                                      endc = dof_handler.end();
+
+  double err_l2 = 0,
+         err_hdiv = 0;
+
+  std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
+
+  for (; cell!=endc; ++cell)
+    {
+      cell->get_dof_indices (local_dof_indices);
+
+      fe_values.reinit (cell);
+
+      // Get function values
+      std::vector<Vector<double> > this_value(n_q_points,
+                                              Vector<double>(n_components));
+      fe_values.get_function_values (solution, this_value);
+
+      // Get values from solution vector (For Trap.Rule)
+      std::vector<std::vector<Tensor<1,3> > >
+      grads_here (n_q_points, std::vector<Tensor<1,3> > (n_components));
+      fe_values.get_function_gradients (solution, grads_here);
+
+      for (unsigned int q_point=0; q_point<n_q_points; ++q_point)
+        {
+          //    double u0 = this_value[q_point](0);
+          //double v0 = this_value[q_point](1);
+
+          double u0 = 0;
+          double v0 = 0;
+          for (unsigned int i = 0; i < dofs_per_cell; ++i)
+            {
+              u0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 0));
+              v0 += (solution (local_dof_indices[i]) *
+                     fe_values.shape_value_component(i, q_point, 1));
+            }
+          u0 -= 1.0;
```

Comment:
what does this do, is this the exact gradient? Can you add a comment or clean it up?